<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from pupper.urdf.xacro              | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="pupper">
  <material name="red">
    <color rgba="1 0 0 1"/>
  </material>
  <material name="yellow">
    <color rgba="1 1 0 1"/>
  </material>
  <material name="blue">
    <color rgba="0 0 1 1"/>
  </material>
  <material name="green">
    <color rgba="0 1 0 1"/>
  </material>
  <material name="blueGreen">
    <color rgba="0 1 1 1"/>
  </material>
  <material name="pink">
    <color rgba="1 0 1 1"/>
  </material>
  <!-- (g*mm^2) -->
  <!-- +++++++++++++++++++++++++++++++ -->
  <!-- MASS PROPERTIES FROM SOLIDWORKS -->
  <!-- +++++++++++++++++++++++++++++++ -->
  <!-- (g) (lumped with all fixed bodies) -->
  <!-- (g) -->
  <!-- (g) -->
  <!-- (g) -->
  <!-- (g) -->
  <!-- (g) -->
  <!-- (g) -->
  <!-- Battery mass -->
  <!-- <xacro:property name = "battery_mass" value = "0.205"/> -->
  <!-- Battery location in x of bottom_PCB frame (m)-->
  <!-- <xacro:property name = "batt_loc_x" value = "-0.028"/> -->
  <!-- ++++++++++++++++++++++++++++++++++++ -->
  <!-- KINEMATIC PROPERTIES FROM SOLIDWORKS -->
  <!-- ++++++++++++++++++++++++++++++++++++ -->
  <!-- Other origins hard coded ... -->
  <!-- +++++++++++++++++++ -->
  <!-- SIMULATION SETTINGS -->
  <!-- +++++++++++++++++++ -->
  <!-- Connect the plugin -->
  <gazebo>
    <plugin filename="libpupper_plugin.so" name="pupper_plugin"/>
  </gazebo>
  <!-- ++++++++++++++++ -->
  <!-- ROBOT START HERE -->
  <!-- ++++++++++++++++ -->
  <!-- Base link of the robot -->
  <link name="bottom_PCB">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.016460000000000002 0.0 0.03753"/>
      <mass value="1.0865"/>
      <inertia ixx="0.00191939864" ixy="2.3571e-07" ixz="1.8848900000000003e-06" iyy="0.00875902112" iyz="1.0990000000000001e-08" izz="0.009498546460000002"/>
    </inertial>
    <visual name="bottom_PCB_visual">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Body/assembly_body_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="bottom_PCB_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Body/assembly_body_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <!-- Connect battery to bottom PCB -->
  <!-- 
    <joint name = "battery_to_body" type = "fixed">
        <parent link = "bottom_PCB"/>
        <child  link = "battery"/>
        <origin xyz = "${batt_loc_x} 0.00 ${batt_z/2}" rpy = "0 0 ${PI/2}"/>
    </joint>
    -->
  <link name="back_left_hub">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.029 0.00031 0.0"/>
      <mass value="0.1098"/>
      <inertia ixx="0.00282357671" ixy="-3.426e-08" ixz="-1.0000000000000001e-11" iyy="1.2321790000000002e-05" iyz="1.0000000000000001e-11" izz="2.7265990000000002e-05"/>
    </inertial>
    <visual name="back_left_hub_visual">
      <material name="blue"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_left_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="back_left_hub_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_left_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="back_left_upper_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0 0.03676 -0.05983"/>
      <mass value="0.14355"/>
      <inertia ixx="0.00017173023" ixy="1.0000000000000001e-11" ixz="-6e-11" iyy="0.00014119183" iyz="4.5763540000000006e-05" izz="4.303383e-05"/>
    </inertial>
    <visual name="back_left_upper_link_visual">
      <material name="yellow"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_left_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="back_left_upper_link_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_left_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="back_left_lower_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.00022 -0.01219 -0.033350000000000005"/>
      <mass value="0.021"/>
      <inertia ixx="4.375377e-05" ixy="1.276e-08" ixz="-3.5976e-07" iyy="0.0028441381500000002" iyz="4.5647000000000006e-07" izz="8.5689e-07"/>
    </inertial>
    <visual>
      <material name="blueGreen"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_lower_link_left_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="back_left_lower_link_collision">
      <pose>0.0 -0.0095 -0.11 0 0 0</pose>
      <geometry>
        <sphere>
          <radius>0.0095</radius>
        </sphere>
      </geometry>
      <!-- <geometry> -->
      <!--     <mesh filename = "${CollisionPath}/Leg/assembly_lower_link_left_simplified.dae" scale = "${meshScale_meters}"/> -->
      <!--  </geometry> -->
    </collision>
  </link>
  <gazebo reference="back_left_lower_link">
    <collision>
      <surface>
        <contact>
          <ode>
            <kp>1e15</kp>
            <!-- 1e15 -->
            <kd>1</kd>
            <!-- 1e13 -->
            <!-- <min_depth>0.01</min_depth> -->
          </ode>
        </contact>
        <friction>
          <ode>
            <mu>0.9</mu>
            <!--floor friction-->
            <mu2>0.9</mu2>
          </ode>
        </friction>
      </surface>
    </collision>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="back_left_hip_joint" type="revolute">
    <parent link="bottom_PCB"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="back_left_hub"/>
    <origin rpy="0 0 0" xyz="-0.122 0.047 0.043500000000000004"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="-1 0 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="back_left_hip_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="back_left_hub">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="back_left_shoulder_joint" type="revolute">
    <parent link="back_left_hub"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="back_left_upper_link"/>
    <origin rpy="0 0 0" xyz="-0.0315 0.0215 0"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="back_left_shoulder_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="back_left_upper_link">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="back_left_elbow_joint" type="revolute">
    <parent link="back_left_upper_link"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="back_left_lower_link"/>
    <origin rpy="0 0 0" xyz="0 0.023 -0.080"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="back_left_elbow_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="back_left_lower_link">
    <self_collide>false</self_collide>
  </gazebo>
  <link name="back_right_hub">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.029 -0.00031 0.0"/>
      <mass value="0.1098"/>
      <inertia ixx="0.00282357671" ixy="-3.426e-08" ixz="-1.0000000000000001e-11" iyy="1.2321790000000002e-05" iyz="-2.0000000000000002e-11" izz="2.7265990000000002e-05"/>
    </inertial>
    <visual name="back_right_hub_visual">
      <material name="blue"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_right_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="back_right_hub_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_right_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="back_right_upper_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0 -0.03676 -0.05983"/>
      <mass value="0.14355"/>
      <inertia ixx="0.00017173023" ixy="1.0000000000000001e-11" ixz="6e-11" iyy="0.00014119183" iyz="-4.5763540000000006e-05" izz="4.303383e-05"/>
    </inertial>
    <visual name="back_right_upper_link_visual">
      <material name="yellow"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_right_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="back_right_upper_link_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_right_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="back_right_lower_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.00022 0.01219 -0.033350000000000005"/>
      <mass value="0.021"/>
      <inertia ixx="4.375305e-05" ixy="-1.274e-08" ixz="-3.6013000000000004e-07" iyy="0.00284413754" iyz="-4.5513e-07" izz="8.568200000000001e-07"/>
    </inertial>
    <visual>
      <material name="blueGreen"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_lower_link_right_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="back_right_lower_link_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_lower_link_right_simplified.dae" scale="1.0 1.0 1.0"/>
        <!-- <mesh filename = "${CollisionPath}/Leg/assembly_lower_link_right_STL.stl" scale = "${meshScale_meters}"/> -->
      </geometry>
    </collision>
  </link>
  <gazebo reference="back_left_lower_link">
    <collision>
      <surface>
        <contact>
          <ode>
            <kp>1e15</kp>
            <!-- 1e15 -->
            <kd>1</kd>
            <!-- 1e13 -->
          </ode>
        </contact>
        <friction>
          <ode>
            <mu>0.9</mu>
            <!--floor friction-->
            <mu2>0.9</mu2>
          </ode>
        </friction>
      </surface>
    </collision>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="back_right_hip_joint" type="revolute">
    <parent link="bottom_PCB"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="back_right_hub"/>
    <origin rpy="0 0 0" xyz="-0.122 -0.047 0.0435"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="-1 0 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="back_right_hip_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="back_right_hub">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="back_right_shoulder_joint" type="revolute">
    <parent link="back_right_hub"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="back_right_upper_link"/>
    <origin rpy="0 0 0" xyz="-0.0315 -0.0215 0"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="back_right_shoulder_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="back_right_upper_link">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="back_right_elbow_joint" type="revolute">
    <parent link="back_right_upper_link"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="back_right_lower_link"/>
    <origin rpy="0 0 0" xyz="0 -0.023 -0.080"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="back_right_elbow_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="back_right_lower_link">
    <self_collide>false</self_collide>
  </gazebo>
  <link name="front_left_hub">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.029 0.00031 0.0"/>
      <mass value="0.1098"/>
      <inertia ixx="0.00282357671" ixy="-3.426e-08" ixz="-1.0000000000000001e-11" iyy="1.2321790000000002e-05" iyz="1.0000000000000001e-11" izz="2.7265990000000002e-05"/>
    </inertial>
    <visual name="front_left_hub_visual">
      <material name="blue"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_left_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="front_left_hub_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_left_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="front_left_upper_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0 0.03676 -0.05983"/>
      <mass value="0.14355"/>
      <inertia ixx="0.00017173023" ixy="1.0000000000000001e-11" ixz="-6e-11" iyy="0.00014119183" iyz="4.5763540000000006e-05" izz="4.303383e-05"/>
    </inertial>
    <visual name="front_left_upper_link_visual">
      <material name="yellow"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_left_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="front_left_upper_link_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_left_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="front_left_lower_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.00022 -0.01219 -0.033350000000000005"/>
      <mass value="0.021"/>
      <inertia ixx="4.375377e-05" ixy="1.276e-08" ixz="-3.5976e-07" iyy="0.0028441381500000002" iyz="4.5647000000000006e-07" izz="8.5689e-07"/>
    </inertial>
    <visual>
      <material name="blueGreen"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_lower_link_left_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="front_left_lower_link_collision">
      <pose>0.0 -0.0095 -0.11 0 0 0</pose>
      <geometry>
        <sphere>
          <radius>0.0095</radius>
        </sphere>
      </geometry>
      <!-- <geometry> -->
      <!--     <mesh filename = "${CollisionPath}/Leg/assembly_lower_link_left_simplified.dae" scale = "${meshScale_meters}"/> -->
      <!--  </geometry> -->
    </collision>
  </link>
  <gazebo reference="front_left_lower_link">
    <collision>
      <surface>
        <contact>
          <ode>
            <kp>1e15</kp>
            <!-- 1e15 -->
            <kd>1</kd>
            <!-- 1e13 -->
            <!-- <min_depth>0.01</min_depth> -->
          </ode>
        </contact>
        <friction>
          <ode>
            <mu>0.9</mu>
            <!--floor friction-->
            <mu2>0.9</mu2>
          </ode>
        </friction>
      </surface>
    </collision>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="front_left_hip_joint" type="revolute">
    <parent link="bottom_PCB"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="front_left_hub"/>
    <origin rpy="0 0 0" xyz="0.078 0.047 0.0435"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="-1 0 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="front_left_hip_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="front_left_hub">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="front_left_shoulder_joint" type="revolute">
    <parent link="front_left_hub"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="front_left_upper_link"/>
    <origin rpy="0 0 0" xyz="-0.0315 0.0215 0"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="front_left_shoulder_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="front_left_upper_link">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="front_left_elbow_joint" type="revolute">
    <parent link="front_left_upper_link"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="front_left_lower_link"/>
    <origin rpy="0 0 0" xyz="0 0.023 -0.080"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="front_left_elbow_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="front_left_lower_link">
    <self_collide>false</self_collide>
  </gazebo>
  <link name="front_right_hub">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.029 -0.00031 0.0"/>
      <mass value="0.1098"/>
      <inertia ixx="0.00282357671" ixy="-3.426e-08" ixz="-1.0000000000000001e-11" iyy="1.2321790000000002e-05" iyz="-2.0000000000000002e-11" izz="2.7265990000000002e-05"/>
    </inertial>
    <visual name="front_right_hub_visual">
      <material name="blue"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_right_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="front_right_hub_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_hip_right_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="front_right_upper_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0.0 -0.03676 -0.05983"/>
      <mass value="0.14355"/>
      <inertia ixx="0.00017173023" ixy="1.0000000000000001e-11" ixz="6e-11" iyy="0.00014119183" iyz="-4.5763540000000006e-05" izz="4.303383e-05"/>
    </inertial>
    <visual name="front_right_upper_link_visual">
      <material name="yellow"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_right_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="front_right_upper_link_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_upper_link_right_simplified.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </collision>
  </link>
  <link name="front_right_lower_link">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.00022 0.01219 -0.033350000000000005"/>
      <mass value="0.021"/>
      <inertia ixx="4.375305e-05" ixy="-1.274e-08" ixz="-3.6013000000000004e-07" iyy="0.00284413754" iyz="-4.5513e-07" izz="8.568200000000001e-07"/>
    </inertial>
    <visual>
      <material name="blueGreen"/>
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_lower_link_right_STL.stl" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
    <collision name="front_right_lower_link_collision">
      <geometry>
        <mesh filename="package://workstation/pupper_description/Pupper_STL/Leg/assembly_lower_link_right_simplified.dae" scale="1.0 1.0 1.0"/>
        <!-- <mesh filename = "${CollisionPath}/Leg/assembly_lower_link_right_STL.stl" scale = "${meshScale_meters}"/> -->
      </geometry>
    </collision>
  </link>
  <gazebo reference="front_left_lower_link">
    <collision>
      <surface>
        <contact>
          <ode>
            <kp>1e15</kp>
            <!-- 1e15 -->
            <kd>1</kd>
            <!-- 1e13 -->
          </ode>
        </contact>
        <friction>
          <ode>
            <mu>0.9</mu>
            <!--floor friction-->
            <mu2>0.9</mu2>
          </ode>
        </friction>
      </surface>
    </collision>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="front_right_hip_joint" type="revolute">
    <parent link="bottom_PCB"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="front_right_hub"/>
    <origin rpy="0 0 0" xyz="0.078 -0.047 0.0435"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="-1 0 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="front_right_hip_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="front_right_hub">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="front_right_shoulder_joint" type="revolute">
    <parent link="front_right_hub"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="front_right_upper_link"/>
    <origin rpy="0 0 0" xyz="-0.0315 -0.0215 0"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="front_right_shoulder_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="front_right_upper_link">
    <self_collide>false</self_collide>
  </gazebo>
  <!-- Attach the child link through a revolute joint -->
  <joint name="front_right_elbow_joint" type="revolute">
    <parent link="front_right_upper_link"/>
    <!-- <parent link = "${name}_motor"/> -->
    <child link="front_right_lower_link"/>
    <origin rpy="0 0 0" xyz="0 -0.023 -0.080"/>
    <!-- transformation from parent origin to child origin -->
    <limit effort="1e6" lower="-1e10" upper="1e10" velocity="1e6"/>
    <dynamics damping="0.001" friction="0.17"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- implicitSpringDamper: prevent crashing when damping is set -->
  <!-- provideFeedback     : allows the joint to publish torque data -->
  <gazebo reference="front_right_elbow_joint">
    <physics>
      <ode>
        <max_force>50</max_force>
      </ode>
    </physics>
    <provideFeedback>true</provideFeedback>
    <implicitSpringDamper>true</implicitSpringDamper>
  </gazebo>
  <gazebo reference="front_right_lower_link">
    <self_collide>false</self_collide>
  </gazebo>
</robot>

